# Deber 2 Módulo Python

## Nombre: Lizeth Moreno

### 1.- Acessos a los datos de felicidad

#### Importo paquetes de python

Para llevar a cabo un análisis de datos y manipulación eficiente de arreglos numéricos en Python, incorporamos las bibliotecas pandas y numpy. Numpy se emplea para la lectura de datos desde archivos, utilizando la función 'np.genfromtxt()', que permite cargar datos tabulares desde una URL especificada, por ejemplo, un archivo CSV. Esta función es especialmente útil para manejar datos mixtos, combinando tanto valores numéricos como cadenas de texto.

Por otro lado, pandas desempeña un papel crucial en la estructuración de datos tabulares, así como en la limpieza y manipulación de datos. Ofrece una integración fluida con numpy, lo que facilita la transición entre las dos bibliotecas. Además, pandas es valioso para llevar a cabo operaciones específicas en series temporales, entre otras funcionalidades esenciales para el análisis de datos. En conjunto, estas dos bibliotecas brindan una sólida base para la exploración y transformación de conjuntos de datos en entornos de programación en Python.

In [3]:
import pandas as pd
import numpy as np

In [17]:
url = 'https://raw.githubusercontent.com/cienciadedatos/datos-de-miercoles/master/datos/2019/2019-08-07/felicidad.csv'
felicidad = np.genfromtxt(url, delimiter=',', skip_header=1, dtype=None, encoding='utf-8')

felicidad


array([('Afghanistán', 2008, 3.7235899 , '7.168690204620361', '0.4506623148918152', '50.79999923706055', '0.7181143164634705', '0.17788857221603394', '0.8816863298416138', '0.5176371932029724', '0.2581954896450043', '0.6120721101760864', '-1.929689645767212', '-1.6550843715667725', 1.7746619 , 'NA', 'NA'),
       ('Afghanistán', 2009, 4.40177822, '7.333789825439453', '0.5523084402084351', '51.20000076293945', '0.6788963675498962', '0.20017842948436737', '0.8500354290008545', '0.5839256048202515', '0.23709242045879364', '0.611545205116272', '-2.0440926551818848', '-1.6350247859954834', 1.7226876 , 'NA', 'NA'),
       ('Afghanistán', 2010, 4.75838089, '7.3866286277771', '0.5390751957893372', '51.599998474121094', '0.6001272201538086', '0.13435254991054535', '0.7067660689353943', '0.6182654500007629', '0.2753238081932068', '0.29935741424560547', '-1.9918100833892822', '-1.6171760559082031', 1.87862182, 'NA', 'NA'),
       ...,
       ('Zimbabue', 2016, 3.7354002 , '7.538829326629639', '0.

In [19]:
type(felicidad)

numpy.ndarray

Una vez cargada la base identificamos que 'felicidad' es un objeto de tipo  'numpy.ndarray' , es decir es un arreglo mutidimensional.

In [20]:
# Otro método de lectura
url2 = 'https://raw.githubusercontent.com/cienciadedatos/datos-de-miercoles/master/datos/2019/2019-08-07/felicidad.csv'

# Leer datos con Pandas
datos_felicidad = pd.read_csv(url2)

# Visualizar los primeros registros
print(datos_felicidad.head())

          pais  anio  escalera_vida   log_pib  soporte_social  \
0  Afghanistán  2008       3.723590  7.168690        0.450662   
1  Afghanistán  2009       4.401778  7.333790        0.552308   
2  Afghanistán  2010       4.758381  7.386629        0.539075   
3  Afghanistán  2011       3.831719  7.415019        0.521104   
4  Afghanistán  2012       3.782938  7.517126        0.520637   

   expectativa_vida  libertad  generosidad  percepcion_corrupcion  \
0         50.799999  0.718114     0.177889               0.881686   
1         51.200001  0.678896     0.200178               0.850035   
2         51.599998  0.600127     0.134353               0.706766   
3         51.919998  0.495901     0.172137               0.731109   
4         52.240002  0.530935     0.244273               0.775620   

   afecto_positivo  afecto_negativo  confianza  calidad_democracia  \
0         0.517637         0.258195   0.612072           -1.929690   
1         0.583926         0.237092   0.611545        

In [34]:
print(datos_felicidad.columns)

Index(['pais', 'anio', 'escalera_vida', 'log_pib', 'soporte_social',
       'expectativa_vida', 'libertad', 'generosidad', 'percepcion_corrupcion',
       'afecto_positivo', 'afecto_negativo', 'confianza', 'calidad_democracia',
       'calidad_entrega', 'de_escalera_pais_anio', 'gini_banco_mundial',
       'gini_banco_mundial_promedio'],
      dtype='object')


### 2.- ¿Cómo calculan la medida de tendencia central y dispersión a nivel global para el primer y último año incluido en el set de datos?

Para determinar el año inicial del análisis, calculamos el valor mínimo de la variable 'year'. De manera análoga, para encontrar el año final, calculamos el valor máximo de la misma variable.

In [27]:
# primer año
primer_anio = datos_felicidad['anio'].min()
primer_anio

2005

In [29]:
ultimo_anio = datos_felicidad['anio'].max()
ultimo_anio 

2018


Ahora procedemos a filtrar los datos de felicidad correspondientes a cada uno de los años. En otras palabras, obtendremos los datos específicamente para el año 2005 y, de manera similar, para el año 2018.

In [31]:
datos_primer_anio = datos_felicidad[datos_felicidad['anio'] == primer_anio]
datos_primer_anio

pais  anio  escalera_vida    log_pib  soporte_social  \
59          Australia  2005       7.340688  10.569470        0.967892   
130           Bélgica  2005       7.262290  10.591697        0.934875   
192            Brasil  2005       6.636771   9.417240        0.882923   
258            Canadá  2005       7.418048  10.608347        0.961552   
386   República Checa  2005       6.439257  10.157401        0.918759   
397         Dinamarca  2005       8.018934  10.704770        0.972372   
440            Egipto  2005       5.167754   8.987387        0.847842   
496           Francia  2005       7.093393  10.505228        0.940338   
532          Alemania  2005       6.619550  10.537519        0.963490   
557            Greece  2005       6.006310  10.294153        0.836539   
624           Hungría  2005       5.193933  10.019172        0.929628   
667              Iran  2005       5.308190   9.656068        0.765978   
713            Italia  2005       6.853784  10.534875        0.928001   
738             Japón  2005       6.515817  10.481733        0.927712   
751          Jordania  2005       6.294660   9.082460        0.920013   
844            Líbano  2005       5.491245   9.414857        0.796278   
978            México  2005       6.580658   9.667230        0.902808   
1065     Países Bajos  2005       7.463979  10.680684        0.947358   
1141         Pakistan  2005       5.224658   8.297400        0.590946   
1218          Polonia  2005       5.587209   9.752288        0.921528   
1245          Rumania  2005       5.048648   9.587583        0.837685   
1281   Arabia Saudita  2005       7.079644  10.726314        0.867819   
1399           España  2005       7.152786  10.401288        0.961043   
1432           Suecia  2005       7.376316  10.626856        0.951470   
1532          Turquía  2005       4.718734   9.699514        0.819936   
1591      Reino Unido  2005       6.983557  10.522983        0.978840   
1642        Venezuela  2005       7.169621   9.612341        0.955278   

      expectativa_vida  libertad  generosidad  percepcion_corrupcion  \
59           71.400002  0.934973          NaN               0.390416   
130          69.900002  0.923843          NaN               0.597554   
192          63.299999  0.882186          NaN               0.744994   
258          71.300003  0.957306     0.244575               0.502681   
386          67.000000  0.865235          NaN               0.900733   
397          69.599998  0.971135          NaN               0.236522   
440          59.700001  0.817362          NaN                    NaN   
496          71.300003  0.894819          NaN               0.687851   
532          70.199997  0.846624          NaN               0.781007   
557          70.500000  0.734172          NaN               0.860563   
624          64.599998  0.696874          NaN               0.902811   
667          62.000000  0.651168          NaN               0.636490   
713          71.900002  0.802195          NaN               0.943912   
738          73.199997  0.867779          NaN               0.698930   
751          63.500000       NaN          NaN               0.669727   
844          64.599998  0.703206          NaN               0.945177   
978          66.199997  0.813745          NaN               0.764249   
1065         70.400002  0.901008          NaN               0.571342   
1141         54.200001  0.629996          NaN               0.844436   
1218         66.300003  0.782473          NaN               0.982931   
1245         64.000000  0.800121          NaN               0.956885   
1281         63.500000       NaN          NaN               0.505149   
1399         71.500000  0.916165          NaN               0.777272   
1432         71.199997  0.964395          NaN                    NaN   
1532         62.599998  0.623115          NaN               0.876999   
1591         69.900002  0.922355          NaN               0.398457   
1642         65.400002  0.83

In [32]:
datos_ultimo_anio = datos_felicidad[datos_felicidad['anio'] == ultimo_anio]
datos_ultimo_anio

pais  anio  escalera_vida   log_pib  soporte_social  \
10    Afghanistán  2018       2.694303  7.494588        0.507516   
21        Albania  2018       5.004403  9.412399        0.683592   
28        Algeria  2018       5.043086  9.557952        0.798651   
45      Argentina  2018       5.792797  9.809972        0.899912   
58        Armenia  2018       5.062449  9.119424        0.814449   
...           ...   ...            ...       ...             ...   
1654    Venezuela  2018       5.005663  9.270281        0.886882   
1667      Vietnam  2018       5.295547  8.783416        0.831945   
1678        Yemen  2018       3.057514       NaN        0.789422   
1690       Zambia  2018       4.041488  8.223958        0.717720   
1703     Zimbabue  2018       3.616480  7.553395        0.775388   

      expectativa_vida  libertad  generosidad  percepcion_corrupcion  \
10           52.599998  0.373536    -0.084888               0.927606   
21           68.699997  0.824212     0.005385               0.899129   
28           65.900002  0.583381    -0.172413               0.758704   
45           68.800003  0.845895    -0.206937               0.855255   
58           66.900002  0.807644    -0.149109               0.676826   
...                ...       ...          ...                    ...   
1654         66.500000  0.610855    -0.176156               0.827560   
1667         67.900002  0.909260    -0.039124               0.808423   
1678         56.700001  0.552726          NaN               0.792587   
1690         55.299999  0.790626     0.036644               0.810731   
1703         55.599998  0.762675    -0.038384               0.844209   

      afecto_positivo  afecto_negativo  confianza  calidad_democracia  \
10           0.424125         0.404904   0.364666                 NaN   
21           0.713300         0.318997   0.435338                 NaN   
28           0.591043         0.292946        NaN                 NaN   
45           0.820310         0.320502   0.261352                 NaN   
58           0.581488         0.454840   0.670828                 NaN   
...               ...              ...        ...                 ...   
1654         0.759221         0.373658   0.260700                 NaN   
1667         0.692222         0.191061        NaN                 NaN   
1678         0.461114         0.314870   0.308151                 NaN   
1690         0.702698         0.350963   0.606715                 NaN   
1703         0.710119         0.211726   0.550508                 NaN   

      calidad_entrega  de_escalera_pais_anio  gini_banco_mundial  \
10                NaN               1.408344                 NaN   
21                NaN               2.640531                 NaN   
28                NaN               1.973943                 NaN   
45                NaN               2.472559                 NaN   
58                NaN               2.102111                 NaN   
...               ...                    ...                 ...   
1654              NaN               2.891930                 NaN   
1667              NaN               1.676448                 NaN   
1678              NaN               2.402008                 NaN   
1690              NaN               2.783419                 NaN   
1703              NaN               2.498696                 NaN   

      gini_banco_mundial_promedio  
10                            NaN  
21                       0.303250  
28                       0.276000  
45                       0.460938  
58                       0.319250  
...                           ...  
1654                     0.497167  
1667                     0.362750  
1678                     0.357000  
1690                     0.527400  
1703                     0.432000  

[136 rows x 17 columns]

Ahora, procedemos a calcular la media para ambos años.

In [38]:
media_primer_anio = datos_primer_anio['escalera_vida'].mean()
media_primer_anio 

6.446164272449635

In [39]:
media_ultimo_anio = datos_ultimo_anio['escalera_vida'].mean()
media_ultimo_anio

5.502134340650895

Con este resultado a nivel mundial podemos concluir que se tuvo una disminución de la felicidad a lo largo de estos años, en promedio los países experimentan una reducción en niveles de felicidad. 

También su desviación estándar.

In [40]:
desviacion_primer_anio = datos_primer_anio['escalera_vida'].std()
desviacion_primer_anio

0.9191426322726484

In [41]:
desviacion_ultimo_anio = datos_ultimo_anio['escalera_vida'].std()
desviacion_ultimo_anio

1.1034612436939357

La dispersión de los datos alrededor de la media nos indica que existe un amento. La variación en los niveles de felicidad entre los países parece haber experimentado un cambio significativo de 2005 a 2018, con una mayor dispersión de los datos en el último año.

### 3. ¿Cómo acceden a la información de la felicidad del primer y último año para los dos países de su elección?

Escogemos los países de análisis. Filtramos los datos para los dos países en el primer y último año.

In [81]:
pais1 = 'México'
print(pais1)
pais2 = 'Rumania'
print(pais2)

México
Rumania


In [ ]:
Año: 2005

In [85]:
datos_pais1_primer_anio = datos_primer_anio[datos_primer_anio['pais'] == pais1]
print(datos_pais1_primer_anio)

       pais  anio  escalera_vida  log_pib  soporte_social  expectativa_vida  \
978  México  2005       6.580658  9.66723        0.902808         66.199997   

     libertad  generosidad  percepcion_corrupcion  afecto_positivo  \
978  0.813745          NaN               0.764249         0.819803   

     afecto_negativo  confianza  calidad_democracia  calidad_entrega  \
978         0.218943   0.426359           -0.124936        -0.088847   

     de_escalera_pais_anio  gini_banco_mundial  gini_banco_mundial_promedio  
978               2.235451               0.489                       0.4698  


In [86]:
datos_pais2_primer_anio = datos_primer_anio[datos_primer_anio['pais'] == pais2]
print(datos_pais2_primer_anio)

         pais  anio  escalera_vida   log_pib  soporte_social  \
1245  Rumania  2005       5.048648  9.587583        0.837685   

      expectativa_vida  libertad  generosidad  percepcion_corrupcion  \
1245              64.0  0.800121          NaN               0.956885   

      afecto_positivo  afecto_negativo  confianza  calidad_democracia  \
1245         0.642016         0.345687   0.272887            0.246735   

      calidad_entrega  de_escalera_pais_anio  gini_banco_mundial  \
1245        -0.113104               2.281933                 NaN   

      gini_banco_mundial_promedio  
1245                       0.3658  


In [87]:
datos_pais1_ultimo_anio = datos_felicidad[(datos_felicidad['pais'] == pais1) & (datos_felicidad['anio'] == ultimo_anio)]
print(datos_pais1_ultimo_anio)

       pais  anio  escalera_vida   log_pib  soporte_social  expectativa_vida  \
990  México  2018       6.549579  9.769919        0.858069         68.300003   

     libertad  generosidad  percepcion_corrupcion  afecto_positivo  \
990    0.8162    -0.181818               0.808638         0.881713   

     afecto_negativo  confianza  calidad_democracia  calidad_entrega  \
990         0.213254   0.294568                 NaN              NaN   

     de_escalera_pais_anio  gini_banco_mundial  gini_banco_mundial_promedio  
990                2.32633                 NaN                       0.4698  


In [88]:
datos_pais2_ultimo_anio = datos_felicidad[(datos_felicidad['pais'] == pais2) & (datos_felicidad['anio'] == ultimo_anio)]
print(datos_pais2_ultimo_anio)

         pais  anio  escalera_vida    log_pib  soporte_social  \
1256  Rumania  2018       6.150879  10.112093         0.81793   

      expectativa_vida  libertad  generosidad  percepcion_corrupcion  \
1256         67.199997   0.84516    -0.220032                0.92117   

      afecto_positivo  afecto_negativo  confianza  calidad_democracia  \
1256         0.735343         0.298454   0.226351                 NaN   

      calidad_entrega  de_escalera_pais_anio  gini_banco_mundial  \
1256              NaN               2.196567                 NaN   

      gini_banco_mundial_promedio  
1256                       0.3658  


Con todos estos datos, ahora podemos comparar el indicador de felicidad de los países seleccionados en sus primeros y últimos años.

Recordando los valores iniciales a nivel mundial, observamos que, en promedio, el nivel de felicidad es el siguiente:

- Primer año (2005): 6.446164272449635
- Último año (2018): 5.502134340650895

Así mismo, las desviaciones estándar correspondientes son:

- Primer año (2005): 0.9191426322726484
- Último año (2018): 1.1034612436939357

Ahora, analizamos los indicadores de felicidad para los dos países seleccionados:

**País 1 - México:**

Felicidad
- Primer año (2005): 6.580658
- Último año (2018): 6.549579

Desviación estándar
- Primer año (2005): 2.235451
- Último año (2018): 2.32633

**País 2 - Rumania:**

Felicidad
- Primer año (2005): 5.048648
- Último año (2018): 6.150879

Desviación estándar
- Primer año (2005): 2.281933
- Último año (2018): 2.196567

**Conclusiones:**

- Para México, el nivel de felicidad se mantuvo relativamente estable a lo largo del tiempo, con una ligera aumento en la desviación estándar en el último año.
- Para Rumania, el indicador de felicidad experimentó un aumento significativo desde su primer año hasta el último. En cuanto a la desviación estándar, muestra una disminución en la variabilidad de los niveles de felicidad en el último año en comparación con el primero.
- Al comparar los dos países, se concluye que México es más feliz tanto en su primer año como en su último año en comparación con Rumania. Además, comparado con el indicador a nivel mundial, México se mantiene cercano al promedio al principio del período, y ambos países son más felices en comparación con el promedio mundial en el último año.
omedio mundial en el último año.